### PR TITLE
[native] Add e2e tests for window value functions

### PR DESCRIPTION
**Pull request by pramodsatya**
Test plan - testFirstValue, testLastValue, testNthValue, testLead, testLag e2e tests

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 1]
```diff
@@ -50,23 +55,58 @@ protected void createTables()
...
+            "RANGE UNBOUNDED PRECEDING",
+            "RANGE CURRENT ROW",
+            "RANGE BETWEEN UNBOUNDED PRECEDING and UNBOUNDED FOLLOWING",
+
+            // Frame clauses in ROWS mode.
+            "ROWS UNBOUNDED PRECEDING",
+            "ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING",
+            "ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING",
+            "ROWS BETWEEN 5 PRECEDING AND CURRENT ROW",
+            "ROWS BETWEEN 5 PRECEDING AND UNBOUNDED FOLLOWING",
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:69]: This clause is not needed.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 2]
```diff
@@ -50,23 +55,58 @@ protected void createTables()
...
+            "ROWS BETWEEN CURRENT ROW AND 5 FOLLOWING",
+            "ROWS BETWEEN UNBOUNDED PRECEDING AND 5 FOLLOWING",
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:71]: UNBOUNDED PRECEDING AND 5 FOLLOWING is not needed.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 3]
```diff
@@ -50,23 +55,58 @@ protected void createTables()
...
+            "ROWS BETWEEN orderkey PRECEDING AND CURRENT ROW",
+            "ROWS BETWEEN orderkey PRECEDING AND UNBOUNDED FOLLOWING",
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:73]: ROWS BETWEEN orderkey PRECEDING AND UNBOUNDED FOLLOWING
and ROWS BETWEEN UNBOUNDED PRECEDING AND orderkey FOLLOWING are not needed.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 4]
```diff
@@ -50,23 +55,58 @@ protected void createTables()
...
+        if (functionType == FunctionType.VALUE) {
+            overClauses.addAll(OVER_CLAUSES_WITHOUT_ORDER_BY);
+        } 
 
         for (String columnProjection : columnProjections) {
             for (String overClause : overClauses) {
-                queries.add(String.format("SELECT %s, %s OVER (%s) AS rnk FROM orders", columnProjection, rankingFunction, overClause));
+                if (functionType == FunctionType.VALUE) {
+                    for (String frameClause : frameClauses) {
+                        queries.add(String.format("SELECT %s, %s OVER (%s %s) AS val FROM orders", columnProjection, function, overClause, frameClause));
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java]: I feel this would affect the readability of the query, but it would send fewer queries to be executed which is better. Grouped 5 function calls into a single Sql query as suggested.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 5]
```diff
@@ -36,38 +41,71 @@ protected void createTables()
...
 
-        for (String columnProjection : columnProjections) {
-            for (String overClause : overClauses) {
-                queries.add(String.format("SELECT %s, %s OVER (%s) AS rnk FROM orders", columnProjection, rankingFunction, overClause));
+        for (String overClause : overClauses) {
+            for (String frameClause : frameClauses) {
+                count++;
+                windowClause += String.format("%s OVER (%s %s)", function, overClause, frameClause);
+                if (count == QUERY_GROUP_SIZE) {
+                    windowClauseList.add(windowClause);
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:91]: Since we are adding to clauseList only after encountering 5 clauses, it means that the last %5 clauses will not get added to the testing if  frameClauses is not a multiple of 5. Please fix.

Comment (pramodsatya) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:91]: Oh right, sorry for missing that, fixed.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 6]
```diff
@@ -36,38 +41,75 @@ protected void createTables()
...
-        List<String> overClauses = orderBy ? OVER_CLAUSES_WITH_ORDER_BY : OVER_CLAUSES_WITHOUT_ORDER_BY;
+        List<String> overClauses = new ArrayList<>(OVER_CLAUSES_WITH_ORDER_BY);
+        List<String> frameClauses = FRAME_CLAUSES;
+        if (functionType == FunctionType.VALUE) {
+            overClauses.addAll(OVER_CLAUSES_WITHOUT_ORDER_BY);
+        }
+        List<String> windowClauseList = new ArrayList<>();
+        String windowClause = new String();
+        int count = 0;
+        final int QUERY_GROUP_SIZE = 5;
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:84]: Nit : Call this FRAMES_PER_QUERY

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 7]
```diff
@@ -115,4 +153,44 @@ public void testRowNumberWithFilter()
         assertQuery("SELECT sum(rn) FROM (SELECT row_number() over() rn, * from orders) WHERE rn = 10");
         assertQuery("SELECT * FROM (SELECT row_number() over(partition by orderstatus order by orderkey) rn, * from orders) WHERE rn = 1");
     }
+
+    @Test
+    public void testFirstValue()
+    {
+        testWindowFunction("first_value(orderkey)", FunctionType.VALUE);
+        testWindowFunction("first_value(orderdate)", FunctionType.VALUE);
+        testWindowFunction("first_value(totalprice)", FunctionType.VALUE);
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:162]: The invocations with total_price are not needed for first_value, last_value and nth_value. Why did you add them ?

Comment (pramodsatya) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java]: The tests for different column types were added to identify any issues that might come up. It is removed now.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 8]
```diff
@@ -36,38 +41,75 @@ protected void createTables()
...
-            for (String overClause : overClauses) {
-                queries.add(String.format("SELECT %s, %s OVER (%s) AS rnk FROM orders", columnProjection, rankingFunction, overClause));
+        for (String overClause : overClauses) {
+            for (String frameClause : frameClauses) {
+                count++;
+                windowClause += String.format("%s OVER (%s %s)", function, overClause, frameClause);
+                if (count == FRAMES_PER_QUERY) {
+                    windowClauseList.add(windowClause);
+                    count = 0;
+                    windowClause = new String();
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:93]: Can we simply reset windowClause = "" ?

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java [hunk 9]
```diff
@@ -27,6 +28,10 @@
 public abstract class AbstractTestNativeWindowQueries
         extends AbstractTestQueryFramework
 {
+    protected enum FunctionType {
+        RANKING, VALUE,
```
Comment (aditi-pandit) [inline presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java:32]: Nit : name RANK

**Review by aditi-pandit — APPROVED** [review]
